macrosynergy/pnl/proxy_pnl_calc.py

```diff
@@ -7,31 +7,174 @@
 import matplotlib.pyplot as plt
 import seaborn as sns
 
-from typing import List, Union, Tuple, Optional
+from typing import List, Union, Tuple, Optional, Dict, Callable
+from numbers import Number
 
-from macrosynergy.management.simulate import make_qdf
-from macrosynergy.management.utils import reduce_df
+from macrosynergy.management.simulate import make_test_df
+from macrosynergy.download.transaction_costs import download_transaction_costs
+from macrosynergy.management.utils import (
+    reduce_df,
+    get_cid,
+    get_xcat,
+    ticker_df_to_qdf,
+    qdf_to_ticker_df,
+)
+from macrosynergy.management.types import QuantamentalDataFrame
+from macrosynergy.pnl.transaction_costs import TransactionCosts
+
+
+def get_diff_index(df_wide: pd.DataFrame) -> pd.Index:
+    df_diff = df_wide.diff(axis=0)
+    change_index = df_diff.index[(df_diff.abs() > 0).any(axis=1)]
+    return change_index
+
+
+def _replace_strs(
+    list_of_strs: List[str], old_str: str, new_str: str = ""
+) -> List[str]:
+    return [_.replace(old_str, new_str) for _ in list_of_strs]
+
+
+def _prep_dfs_for_pnl_calcs(
+    df: QuantamentalDataFrame,
+    spos: str,
+    rstring: str,
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, List[pd.Timestamp]]:
+    df_wide = qdf_to_ticker_df(df=df.copy())
+
+    # Filter
+    u_tickers: List[str] = list(set(df["cid"] + "_" + df["xcat"]))
+    filt_xrs: List[str] = [tx for tx in u_tickers if tx.endswith(rstring)]
+    # do xrs contain `spos` at this point?
+    filt_pos = [tx for tx in u_tickers if tx.endswith(spos)]
+
+    assert set(_replace_strs(filt_xrs, rstring)) == set(
+        _replace_strs(filt_pos, f"_{spos}")
+    )
+
+    # Pivot the dataframes
+    pivot_returns: pd.DataFrame = df_wide.loc[:, filt_xrs]
+    pivot_pos: pd.DataFrame = df_wide.loc[:, filt_pos]
+
+    # warn about NAs
+    dfx: pd.DataFrame
+    for dfx, dfname in [(pivot_returns, "returns"), (pivot_pos, "positions")]:
+        # for each column warns for dates of nas
+        for col in dfx.columns:
+            nas_idx = dfx[col].loc[dfx[col].isna()]
+            if not nas_idx.empty:
+                print(
+                    f"Warning: Series {col} has NAs at the following dates: {nas_idx.index}"
+                )
+
+    # Get the diff index for positions
+    pos_diff_index: pd.DatetimeIndex = get_diff_index(df_wide=pivot_pos)
```

Comment:
How is the black-listing handled in positions? Are we setting intra-periods N/A to zero if black-listed? We need to make sure only to rebalance for N/A when we go from N/A (or zero) to a value.

Reply:
In the actual entry point `proxy_pnl_calc`

Reply:
(black-listing was just an example of where I think we might have N/A's).